db/db_impl_compaction_flush.cc

```diff
@@ -1144,6 +1154,67 @@ Status DBImpl::FlushMemTable(ColumnFamilyData* cfd,
   return s;
 }
 
+Status DBImpl::FlushMemTables(const FlushOptions& flush_options,
+    FlushReason flush_reason, bool writes_stopped) {
+  Status s;
+  bool schedule_flush = false;
+  GroupFlushRequest flush_req;
+  {
+    WriteContext context;
+    InstrumentedMutexLock guard_lock(&mutex_);
+
+    WriteThread::Writer w;
+    if (!writes_stopped) {
+      write_thread_.EnterUnbatched(&w, &mutex_);
+    }
+
+    for (auto cfd : *versions_->GetColumnFamilySet()) {
+      if (cfd->IsDropped() ||
+          (cfd->imm()->NumNotFlushed() == 0 && cfd->mem()->IsEmpty() &&
+          cached_recoverable_state_empty_.load())) {
+        continue;
+      }
+      s = SwitchMemtable(cfd, &context);
```

Comment:
Since you mentioned this issue twice, I'm going to respond here to both comments. I think this is a good point. Even though `SwitchMemtable` is already serialized by the write queue mechanism, it still needs to acquire the db mutex to protect shared data structures that can be modified outside RocksDB write path. Therefore, RocksDB needs to switch memtables for all column families before initiating atomic flush.

Reply:
We discussed this offline. According to the thread-safety property of `ColumnFamilySet` during iteration, original approach will be safe as long as we use `Ref`/`Unref` column family.